<style  type="text/css">
cool {
  width: 100px;
  height: 100px;
  background-color: red;
  position: relative;
  -webkit-animation-name: example; /* Safari 4.0 - 8.0 */
  -webkit-animation-duration: 4s; /* Safari 4.0 - 8.0 */
  -webkit-animation-iteration-count: infinite; /* Safari 4.0 - 8.0 */
  animation-name: example;
  animation-duration: 4s;
  animation-iteration-count: infinite;
}

loud {
    font-size: 20pt;
    }

/* Safari 4.0 - 8.0 */
@-webkit-keyframes example {
  0%   {background-color:red; left:0px; top:0px;}
  25%  {background-color:yellow; left:200px; top:0px;}
  50%  {background-color:blue; left:200px; top:200px;}
  75%  {background-color:green; left:0px; top:200px;}
  100% {background-color:red; left:0px; top:0px;}
}

/* Standard syntax */
@keyframes example {
  0%   {background-color:red; left:0px; top:0px;}
  25%  {background-color:yellow; left:200px; top:0px;}
  50%  {background-color:blue; left:200px; top:200px;}
  75%  {background-color:green; left:0px; top:200px;}
  100% {background-color:red; left:0px; top:0px;}
}
</style>

<b><center>
<span style="font-size: 24pt; line-height: 1.2">
COMS W4111: Introduction to Databases<br>
Spring 2024, Sections 002/V02
</span>

</center></b>
<br>
<p>
<i><center>
<span style="font-size: 20pt; line-height: 1.2">
Homework 4<br>
</span>
</center></i>


# Introduction

- This notebook contains HW4. **Both Programming and Nonprogramming tracks should complete this homework.**
- You will submit **PDF and ZIP files** for this assignment. Gradescope will have two separate assignments for these.
- For the PDF:
  - The most reliable way to save as PDF is to go to your browser's menu bar and click `File -> Print`. Switch the orientation to landscape mode, and hit save.
  - **MAKE SURE ALL YOUR WORK (CODE AND SCREENSHOTS) IS VISIBLE ON THE PDF. YOU WILL NOT GET CREDIT IF ANYTHING IS CUT OFF.** Reach out for troubleshooting.
  - **MAKE SURE YOU DON'T SUBMIT A SINGLE PAGE PDF.** Your PDF should have multiple pages.
- For the ZIP:
  - Zip a folder containing this notebook and any screenshots.
  - You may delete any unnecessary files, such as caches.


# Setup


In [6]:
%load_ext sql
%sql mysql+pymysql://root:dbuserdbuser@localhost

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [7]:
import sys

!{sys.executable} -m pip install --upgrade pymongo
!{sys.executable} -m pip install --upgrade neo4j


[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: pip install --upgrade pip


- If you get warnings below, try restarting your kernel


In [8]:
import neo4j
import pandas
import pymongo

# TODO: Fill in with your Mongo URL
mongo_url = "mongodb+srv://dbuser:dbuserdbuser@hw4.sdegwse.mongodb.net/?retryWrites=true&w=majority&appName=hw4"
mongo_client = pymongo.MongoClient(mongo_url)

# TODO: Fill in with your Neo4j credentials
neo4j_url = "neo4j+s://4661e20a.databases.neo4j.io"
neo4j_password = "valu28i6wRhaRDB22hyrefkWv9dCvja1vo2D_UlOtl0"
# username is always "neo4j"
graph = neo4j.GraphDatabase.driver(neo4j_url, auth=("neo4j", neo4j_password))
graph.verify_connectivity()

<hr style="height:2px;">


# Written Questions

- As usual, do not bloviate


## W1

Explain the following concepts:

1. Clustering index
2. Nonclustering index
3. Sparse index
4. Dense index

#### Answer

1. Clustering index: the order of the rows in a relational database is sequential
2. Nonclustering index: where the order is something other than the order of the file.
3. Sparse index: contains index records for only some search-key values.
4. Dense index: Index record appears for every search-key value in the file.


## W2

Explain why nonclustering indexes must be dense.

_Answer_

If the records do not have some sort of known order, there's no way to efficiently find them by going to the "nearest" known key-value and then search sequentially until found, which is the approach of a sparse index. Sparse indicies rely on that sequential order, so nonclustering indicies must be dense.


## W3

Suppose that, in a table containing information about Columbia classes, the columns `class_code`, `semester`, and `year` are queried frequently **individually**. Would putting a composite index on `(class_code, semester, year)` be a good idea? Why or why not?

_Answer_

Putting a composite index on `(class_code, semester, year)` would not be a good idea, because a composite key would not cover a query on `semester` alone, for instance.


## W4

Explain the following concepts:

1. Hash index
2. B+ tree index

_Answer_

A hash index is an index which is calculated using a hash function, typically in constant time. Assuming there are no collisions, the hash of a search key specifies which block a record should be stored in.

A B+ tree index is a type of self-balancing, sorted tree structure. To minimize the number of I/O operations for search, the fanout for B+ trees is large. The leaf nodes are the only nodes that actually store data, and leaf nodes are linked to reduce traversal time.


## W5

Give one advantage and one disadvantage of hash indexes compared to B+ tree indexes.

Advantage: hash index access is O(1), which is unbeatable.
Disadvantage: Hash indexes are not efficient for range queries (e.g., finding all entries between two values). They are designed for exact match queries.


## W6

Explain the role of the buffer in a DBMS. Why doesn't the DBMS simply load the entire database in its buffer?

_Answer_

The buffer minimizes the number of disk accesses by storing copies of what's on the disk in the buffer. The DBMS doesn't load the entire database because except where the DB is very small, it wouldn't all fit in memory. Even if it did, RAM is way more expensive than disk storage, so it's impractical.


## W7

Explain the following concepts as they relate to buffer replacement policies:

1. Clairvoyant algorithm
2. Least recently used strategy
3. Most recently used strategy
4. Clock algorithm

_Answer_

1. Clairvoyant algorithm: the policy that says to always disregard the information that will not be needed for the longest time in the future. This is a theoretical benchmark.
2. Least recently used strategy: Tracks page access in a buffer and replaces the page that has not been accessed for the longest time. It assumes that pages accessed recently will likely be accessed again soon.
3. Most recently used strategy: replaces the page that was most recently accessed. It operates under the assumption that the most recently used page is less likely to be used again soon.
4. Clock algorithm: less-expensive approximation of the LRU strategy. Rotate through pages until we find one that hasn’t been referenced since the last time we checked it.


## W8

NoSQL databases have become increasingly popular for applications. List 3 benefits of using NoSQL databases over SQL ones.

_Answer_

1. Scalability: NoSQL databases can easily scale horizontally to handle large volumes of data, making them suitable for big data and high-traffic applications.
2. Flexibility: They allow for a flexible, schema-less data model that accommodates various data types and structures, ideal for rapidly changing data.
3. Performance: NoSQL databases often provide faster performance write operations.


## W9

Explain the concept between impedance mismatch and how it relates to SQL vs. NoSQL databases.

_Answer_
Object oriented programming essentially deals with directed graphs, whereas relational databased deal fundamentally with tuples, creating an impedance mismatch. By using NoSQL databases, developers can often store and manipulate data in a way that is more natural to their application's programming model.


## W10

The relationship between students and courses is many-to-many. Due to its emphasis on atomicity, modeling this relationship in a relational database would require an associative entity. Explain how this relationship could be modeled in

1. A document database, such as MongoDB
2. A graph database, such as Neo4j

_Answer_

#### MongoDB

One way to do this in MongoDB is to use embedded documents, where you embed an array of documents representing the related data directly into the parent document.

#### Neo4j

Because relationships are baked into the definition of the graph database, we can directly link `student` nodes to the related `course` nodes and vice-versa.


<hr style="height:2px;">


# MongoDB

- The cell below creates a database `w4111`, then a collection `episodes` inside `w4111`. It then inserts GoT episode data into the collection.


In [9]:
import json

with open("episodes.json") as f:
    data = json.load(f)

episodes = mongo_client["w4111"]["episodes"]
episodes.drop()
episodes.insert_many(data)
print("Successfully inserted episode data")

Successfully inserted episode data


## M1

- Write and execute a query that shows episodes and the number of scenes they contain
- Your aggregation should have the following attributes:
  - `episodeTitle`
  - `seasonNum`
  - `episodeNum`
  - `numScenes`, which is the length of the episode's `scenes` array
- Order your output on `numScenes` descending, and only keep episodes with more than 100 scenes


In [10]:
res = episodes.aggregate(
    [
        {"$addFields": {"numScenes": {"$size": "$scenes"}}},
        {"$match": {"numScenes": {"$gt": 100}}},
        {
            "$project": {
                "_id": 0,
                "episodeTitle": 1,
                "seasonNum": 1,
                "episodeNum": 1,
                "numScenes": 1,
            }
        },
        {"$sort": {"numScenes": -1}},
    ]
)
pandas.DataFrame(list(res))

seasonNum  episodeNum             episodeTitle  numScenes
0          8           3           The Long Night        292
1          8           5                The Bells        220
2          2           9               Blackwater        133
3          8           4   The Last of the Starks        113
4          7           7  The Dragon and the Wolf        104

## M2

- Write and execute a query that shows the first three episodes for each season
- Your aggregation should have the following attributes:
  - `seasonNum`
  - `firstThreeEpisodes`, which is an array that contains the titles of the first, second, and third episodes (in that order) of the season
- Order your output on `seasonNum` ascending
  - It's okay if the `firstThreeEpisodes` column is a bit truncated by the dataframe


In [11]:
res = episodes.aggregate(
    [
        {"$sort": {"seasonNum": 1, "episodeNum": 1}},
        {
            "$group": {
                "_id": "$seasonNum",
                "firstThreeEpisodes": {"$push": "$episodeTitle"},
            }
        },
        {
            "$project": {
                "seasonNum": "$_id",
                "_id": 0,
                "firstThreeEpisodes": {"$slice": ["$firstThreeEpisodes", 3]},
            }
        },
        {"$sort": {"seasonNum": 1}},
    ]
)

pandas.DataFrame(list(res))

seasonNum                                 firstThreeEpisodes
0          1       [Winter Is Coming, The Kingsroad, Lord Snow]
1          2  [The North Remembers, The Night Lands, What Is...
2          3  [Valar Dohaeris, Dark Wings, Dark Words, Walk ...
3          4  [Two Swords, The Lion and the Rose, Breaker of...
4          5  [The Wars to Come, The House of Black and Whit...
5          6                 [The Red Woman, Home, Oathbreaker]
6          7      [Dragonstone, Stormborn, The Queen's Justice]
7          8  [Winterfell, A Knight of the Seven Kingdoms, T...

## M3

- Write and execute a query that shows statistics about each season
- Your aggregation should have the following attributes:
  - `seasonNum`
  - `numEpisodes`, which is the number of episodes in the season
  - `startDate`, which is the earliest air date associated with an episode in the season
  - `endDate`, which is the latest air date associated with an episode in the season
  - `shortestEpisodeLength`
  - `longestEpisodeLength`
    - The length of an episode is the greatest `sceneEnd` value in the episode's `scenes` array
- Order your output on `seasonNum` ascending


In [12]:
res = episodes.aggregate(
    [
        {"$addFields": {"maxSceneEnd": {"$max": "$scenes.sceneEnd"}}},
        {
            "$group": {
                "_id": "$seasonNum",
                "numEpisodes": {"$sum": 1},
                "startDate": {"$min": "$episodeAirDate"},
                "endDate": {"$max": "$episodeAirDate"},
                "shortestEpisodeLength": {"$min": "$maxSceneEnd"},
                "longestEpisodeLength": {"$max": "$maxSceneEnd"},
            }
        },
        {
            "$project": {
                "_id": 0,
                "seasonNum": "$_id",
                "numEpisodes": "$numEpisodes",
                "startDate": "$startDate",
                "endDate": "$endDate",
                "shortestEpisodeLength": "$shortestEpisodeLength",
                "longestEpisodeLength": "$longestEpisodeLength",
            }
        },
        {"$sort": {"seasonNum": 1}},
    ]
)

pandas.DataFrame(list(res))

seasonNum  numEpisodes   startDate     endDate shortestEpisodeLength  \
0          1           10  2011-04-17  2011-06-19               0:51:30   
1          2           10  2012-04-01  2012-06-03               0:49:18   
2          3           10  2013-03-31  2013-06-09               0:49:25   
3          4           10  2014-04-06  2014-06-15               0:49:19   
4          5           10  2015-03-29  2015-06-14               0:50:32   
5          6           10  2016-04-24  2016-06-26               0:51:52   
6          7            7  2017-07-16  2017-08-27               0:50:05   
7          8            6  2019-04-14  2019-05-19               0:54:29   

  longestEpisodeLength  
0              1:00:57  
1              1:02:04  
2              1:01:20  
3              1:04:49  
4              1:02:01  
5              1:10:14  
6              1:21:10  
7              1:21:37

## M4

- Write and execute a query that shows sublocations and the scenes they appear in
- Your aggregation should have the following attributes:
  - `subLocation`
  - `totalScenes`, which is the number of scenes that are set in the sublocation
  - `firstSeasonNum`
  - `firstEpisodeNum`
    - `(firstSeasonNum, firstEpisodeNum)` identifies the first episode that the sublocation appears in
  - `lastSeasonNum`
  - `lastEpisodeNum`
    - `(lastSeasonNum, lastEpisodeNum)` identifies the last episode that the sublocation appears in
- Order your output on `totalScenes` descending, and only keep the sublocations with more than 50 scenes


In [13]:
res = episodes.aggregate(
    [
        {"$unwind": "$scenes"},  # Flatten the scenes array to process individual scenes
        {
            "$group": {
                "_id": "$scenes.subLocation",  # Group by subLocation
                "totalScenes": {"$sum": 1},  # Count the total number of scenes
                "firstSeasonNum": {"$min": "$seasonNum"},  # Get the first season number
                "firstEpisodeNum": {
                    "$min": "$episodeNum"
                },  # Get the first episode number
                "lastSeasonNum": {"$max": "$seasonNum"},  # Get the last season number
                "lastEpisodeNum": {
                    "$max": "$episodeNum"
                },  # Get the last episode number
            }
        },
        {"$match": {"totalScenes": {"$gt": 50}}},
        {
            "$project": {
                "_id": 0,
                "subLocation": "$_id",
                "totalScenes": "$totalScenes",
                "firstSeasonNum": "$firstSeasonNum",
                "firstEpisodeNum": "$firstEpisodeNum",
                "lastSeasonNum": "$lastSeasonNum",
                "lastEpisodeNum": "$lastEpisodeNum",
            }
        },
        {"$sort": {"totalScenes": -1}},
    ]
)

pandas.DataFrame(list(res))

subLocation  totalScenes  firstSeasonNum  firstEpisodeNum  \
0       King's Landing         1094               1                1   
1           Winterfell          734               1                1   
2                 None          494               1                1   
3         Castle Black          267               1                1   
4          Dragonstone          142               2                1   
5   The Haunted Forest           77               1                1   
6   Outside Winterfell           69               1                1   
7       Craster's Keep           66               2                1   
8             The Wall           60               2                1   
9            The Twins           57               1                1   
10     Blackwater Rush           56               7                4   
11      Blackwater Bay           53               2                1   

    lastSeasonNum  lastEpisodeNum  
0               8              10  
1               8              10  
2               8              10  
3               8              10  
4               8              10  
5               8              10  
6               8              10  
7               4               5  
8               8              10  
9               7              10  
10              7               5  
11              8               9

<hr style="height:2px;">


# Neo4j

- The cell below creates nodes and relationships that model movies and the people involved in them


In [14]:
with open("movies.txt") as f:
    queries = str(f.read())

graph.execute_query("match (p:Person), (m:Movie) detach delete p, m")
graph.execute_query(queries)
print("Successfully inserted movie data")

Successfully inserted movie data


## N1

- Write and execute a cypher that shows actors and the number of movies they appear in
  - You should focus only on the `ACTED_IN` relationship, no other relationship
- Your output should have the following attributes:
  - `name`, which is the name of the actor
  - `num_movies`
- Order your output on `num_movies` descending, and only keep actors who have acted in 4 or more movies


In [15]:
res = graph.execute_query(
    """
    MATCH (a:Person)-[r:ACTED_IN]->(m:Movie)
    WITH a.name AS name, COUNT(m) AS num_movies
    WHERE num_movies >= 4
    RETURN name, num_movies
    ORDER BY num_movies DESC
"""
)

pandas.DataFrame([dict(r) for r in res.records])

name  num_movies
0         Tom Hanks          12
1      Keanu Reeves           7
2      Hugo Weaving           5
3    Jack Nicholson           5
4          Meg Ryan           5
5  Cuba Gooding Jr.           4

## N2

- Write and execute a cypher that shows people and movies they either acted in or directed
- Your output should have the following attributes:
  - `name`, which is the name of the person
  - `directed_movies`, which is an array of titles of movies that the person directed
  - `acted_in_movies`, which is an array of titles of movies that the person acted in
- Order your output on `name` ascending, and only keep people that have directed at least one movie **and** acted in at least one movie (i.e., there should be no empty arrays. Arrays with one element are fine.)


In [16]:
res = graph.execute_query(
    """
    MATCH (p:Person)-[:DIRECTED]->(d:Movie), (p)-[:ACTED_IN]->(a:Movie)
    WITH p.name AS name, 
         COLLECT(DISTINCT d.title) AS directed_movies, 
         COLLECT(DISTINCT a.title) AS acted_in_movies
    WHERE SIZE(directed_movies) > 0 AND SIZE(acted_in_movies) > 0
    RETURN name, directed_movies, acted_in_movies
    ORDER BY name
"""
)

pandas.DataFrame([dict(r) for r in res.records])

name                   directed_movies  \
0  Clint Eastwood                      [Unforgiven]   
1    Danny DeVito                           [Hoffa]   
2  James Marshall  [V for Vendetta, Ninja Assassin]   
3       Tom Hanks               [That Thing You Do]   
4   Werner Herzog                      [RescueDawn]   

                                     acted_in_movies  
0                                       [Unforgiven]  
1           [Hoffa, One Flew Over the Cuckoo's Nest]  
2                                   [A Few Good Men]  
3  [You've Got Mail, Sleepless in Seattle, Joe Ve...  
4                             [What Dreams May Come]

## N3

- Write and execute a cypher that shows people and movies they both acted in and directed
- Your output should have the following attributes:
  - `name`, which is the name of the person
  - `acted_in_and_directed_movies`, which is an array of titles of movies that the person both acted in and directed
- Order your output on `name` ascending, and only keep people that have acted in at least one movie that they directed (i.e., there should be no empty arrays. Arrays with one element are fine.)


In [17]:
res = graph.execute_query(
    """
    MATCH (p:Person)-[:ACTED_IN]->(m:Movie)
    WHERE (p)-[:DIRECTED]->(m)
    WITH p.name AS name, COLLECT(m.title) AS acted_in_and_directed_movies
    WHERE SIZE(acted_in_and_directed_movies) > 0
    RETURN name, acted_in_and_directed_movies
    ORDER BY name
"""
)

pandas.DataFrame([dict(r) for r in res.records])

name acted_in_and_directed_movies
0  Clint Eastwood                 [Unforgiven]
1    Danny DeVito                      [Hoffa]
2       Tom Hanks          [That Thing You Do]

## N4

- Write and execute a cypher that shows pairs of people and how closely connected they are
- Your output should have the following attributes:
  - `person_1_name`, which is the name of the first person in the pair
  - `person_2_name`, which is the name of the second person in the pair
  - `num_people_between`, which is the number of people (including the pair itself) separating the pair. You should use the `shortestPath` function to compute this.
- To prevent duplicates in your output, you should only keep rows where `person_1_name < person_2_name`
- Order your output on `(person_1_name, person_2_name)`, and only keep rows where `num_people_between > 5`

- As an example, you should get the following row in your output:

| person_1_name | person_2_name | num_people_between |
| ------------- | ------------- | ------------------ |
| Billy Crystal | Paul Blythe   | 6                  |

- The shortest path between Billy Crystal and Paul Blythe is shown below
  - `num_people_between` is 6 because there are 6 nodes marked as `Person` (including Billy's and Paul's nodes)

<img src="./N4-example.png">


In [18]:
res = graph.execute_query(
    """
    MATCH path = shortestPath((p1:Person)-[*]-(p2:Person))
    WHERE p1.name < p2.name
    WITH p1, p2, nodes(path) AS path_nodes
    WITH p1.name AS person_1_name, p2.name AS person_2_name, 
        SIZE([node IN path_nodes WHERE 'Person' IN LABELS(node)]) AS num_people_between
    WHERE num_people_between > 5
    RETURN person_1_name, person_2_name, num_people_between
    ORDER BY person_1_name, person_2_name

    """
)

pandas.DataFrame([dict(r) for r in res.records])

person_1_name   person_2_name  num_people_between
0    Billy Crystal     Paul Blythe                   6
1      Bruno Kirby     Paul Blythe                   6
2    Carrie Fisher     Paul Blythe                   6
3   Christian Bale      Dina Meyer                   6
4   Christian Bale           Ice-T                   6
5   Christian Bale     Paul Blythe                   6
6   Christian Bale    Robert Longo                   6
7   Christian Bale  Takeshi Kitano                   6
8      Ethan Hawke     Paul Blythe                   6
9      Jan de Bont     Paul Blythe                   6
10     Paul Blythe     Scott Hicks                   6
11     Paul Blythe    Zach Grenier                   6

<hr style="height:2px;">


# SQL To NoSQL

- You will move relational data to document and graph databases
  - **You will do your modeling in Python. You shouldn't be writing any SQL.**
- You will be using the `classicmodels` database for this section. You may want to drop the database and re-run the SQL script (included in the directory) to ensure you have the right data.
  - You will be modeling customers and the products they ordered


## MongoDB: Customers

- For the document database, you will create two collections: `customers` and `products`
- `customers` will contain customer information as well as all the orders they've placed
- You will use `customer_orders_all_df` to create your `customers` collection


In [19]:
%%sql

customer_orders_all <<

select
    c.customerNumber, c.customerName, c.country, o.orderNumber,
    o.orderDate, od.productCode, od.quantityOrdered, od.priceEach
from classicmodels.orders o
    join classicmodels.customers c using (customerNumber)
    join classicmodels.orderdetails od using (orderNumber);

 * mysql+pymysql://root:***@localhost
2996 rows affected.
Returning data to local variable customer_orders_all


In [20]:
customer_orders_all_df = customer_orders_all.DataFrame()
customer_orders_all_df.head(10)

customerNumber        customerName        country  orderNumber   orderDate  \
0             103   Atelier graphique         France        10123  2003-05-20   
1             103   Atelier graphique         France        10123  2003-05-20   
2             103   Atelier graphique         France        10123  2003-05-20   
3             103   Atelier graphique         France        10123  2003-05-20   
4             103   Atelier graphique         France        10298  2004-09-27   
5             103   Atelier graphique         France        10298  2004-09-27   
6             103   Atelier graphique         France        10345  2004-11-25   
7             112  Signal Gift Stores  United States        10124  2003-05-21   
8             112  Signal Gift Stores  United States        10124  2003-05-21   
9             112  Signal Gift Stores  United States        10124  2003-05-21   

  productCode  quantityOrdered priceEach  
0    S18_1589               26    120.71  
1    S18_2870               46    114.84  
2    S18_3685               34    117.26  
3    S24_1628               50     43.27  
4    S10_2016               39    105.86  
5    S18_2625               32     60.57  
6    S24_2022               43     38.98  
7    S18_1749               21    153.00  
8    S18_2248               42     58.12  
9    S18_2325               42    111.87

- Below is an example of how a customer and their orders are stored in MySQL, and how the document should look like in MongoDB
- The document should have the following attributes:
  - `customerNumber`
  - `customerName`
  - `country`
  - `orders`, which is a list of objects. Each object represents one order
    - `orderNumber`
    - `orderDate`
    - `orderContents`, which is a list of objects. Each object represents one product in the order
      - `productCode`
      - `quantityOrdered`
      - `priceEach`

**MySQL relation:**

|     | customerNumber | customerName      | country | orderNumber | orderDate  | productCode | quantityOrdered | priceEach |
| --: | -------------: | :---------------- | :------ | ----------: | :--------- | :---------- | --------------: | --------: |
|   0 |            103 | Atelier graphique | France  |       10123 | 2003-05-20 | S18_1589    |              26 |    120.71 |
|   1 |            103 | Atelier graphique | France  |       10123 | 2003-05-20 | S18_2870    |              46 |    114.84 |
|   2 |            103 | Atelier graphique | France  |       10123 | 2003-05-20 | S18_3685    |              34 |    117.26 |
|   3 |            103 | Atelier graphique | France  |       10123 | 2003-05-20 | S24_1628    |              50 |     43.27 |
|   4 |            103 | Atelier graphique | France  |       10298 | 2004-09-27 | S10_2016    |              39 |    105.86 |
|   5 |            103 | Atelier graphique | France  |       10298 | 2004-09-27 | S18_2625    |              32 |     60.57 |
|   6 |            103 | Atelier graphique | France  |       10345 | 2004-11-25 | S24_2022    |              43 |     38.98 |

**MongoDB document:**

```
{
    customerNumber: 103
    customerName: "Atelier graphique",
    country: "France",
    orders: [
        {
            orderNumber: 10123,
            orderDate: "2003-05-20",
            orderContents: [
                {
                    productCode: "S18_1589",
                    quantityOrdered: 26,
                    priceEach: "120.71"
                },
                {
                    productCode: "S18_2870",
                    quantityOrdered: 46,
                    priceEach: "114.84"
                },
                {
                    productCode: "S18_3685",
                    quantityOrdered: 34,
                    priceEach: "117.26"
                },
                {
                    productCode: "S24_1628",
                    quantityOrdered: 50,
                    priceEach: "43.27"
                }
            ]
        },
        {
            orderNumber: 10298,
            orderDate: "2004-09-27",
            orderContents: [
                {
                    productCode: "S10_2016",
                    quantityOrdered: 39,
                    priceEach: "105.86"
                },
                {
                    productCode: "S18_2625",
                    quantityOrdered: 32,
                    priceEach: "60.57"
                }
            ]
        },
        {
            orderNumber: 10345,
            orderDate: "2004-11-25",
            orderContents: [
                {
                    productCode: "S24_2022",
                    quantityOrdered: 43,
                    priceEach: "38.98"
                }
            ]
        }
    ]
}
```


In [21]:
# TODO: Create a list of dicts. Each dict represents one customer.


"""
Tips:

    To iterate through dataframe:
    
        for _, r in customer_orders_all_df.iterrows():
            r = dict(r)
            Access fields like r['customerName'], r['country'], ...
            
    
    The orderDate and priceEach fields are stored as datetime.date and Decimal
    objects in the dataframe. These types are not compatible with the pymongo API.
    You can convert them to strings by calling str(r['orderDate']) and str(r['priceEach']). 
    Alternatively, you can look into the datetime.datetime and bson.decimal128.Decimal128 
    objects, which are supported by pymongo.

"""


customer_orders_all_df["orderDate"] = customer_orders_all_df["orderDate"].astype(str)
customer_orders_all_df["priceEach"] = customer_orders_all_df["priceEach"].astype(str)


result = (
    customer_orders_all_df.groupby(
        ["customerNumber", "customerName", "country", "orderNumber", "orderDate"]
    )
    .apply(
        lambda x: x[["productCode", "quantityOrdered", "priceEach"]].to_dict("records")
    )
    .reset_index()
    .rename(columns={0: "orderContents"})
)

customer_list = (
    result.groupby(["customerNumber", "customerName", "country"])
    .apply(
        lambda x: x[["orderNumber", "orderDate", "orderContents"]].to_dict("records")
    )
    .reset_index()
    .rename(columns={0: "orders"})
    .to_dict("records")
)

In [22]:
def insert_customers(d):
    mongo_client["w4111"]["customers"].drop()
    mongo_client["w4111"]["customers"].insert_many(d)


# TODO: Put the name of your list of dicts below
insert_customers(customer_list)
print("Successfully inserted customer data")

Successfully inserted customer data


## MongoDB: Products

- To create the `products` collection, you will use `products_all_df`
- A document in `products` simply contains product information, as shown below

```
{
    productCode: "S10_1678",
    productName: "1969 Harley Davidson Ultimate Chopper",
    productVendor: "Min Lin Diecast"
}
```


In [23]:
%%sql

products_all <<

select productCode, productName, productVendor
from classicmodels.products;

 * mysql+pymysql://root:***@localhost
110 rows affected.
Returning data to local variable products_all


In [24]:
products_all_df = products_all.DataFrame()
products_all_df.head(10)

productCode                            productName  \
0    S10_1678  1969 Harley Davidson Ultimate Chopper   
1    S10_1949               1952 Alpine Renault 1300   
2    S10_2016                  1996 Moto Guzzi 1100i   
3    S10_4698   2003 Harley-Davidson Eagle Drag Bike   
4    S10_4757                    1972 Alfa Romeo GTA   
5    S10_4962                 1962 LanciaA Delta 16V   
6    S12_1099                      1968 Ford Mustang   
7    S12_1108                      2001 Ferrari Enzo   
8    S12_1666                         1958 Setra Bus   
9    S12_2823                       2002 Suzuki XREO   

               productVendor  
0            Min Lin Diecast  
1    Classic Metal Creations  
2   Highway 66 Mini Classics  
3          Red Start Diecast  
4    Motor City Art Classics  
5        Second Gear Diecast  
6      Autoart Studio Design  
7        Second Gear Diecast  
8  Welly Diecast Productions  
9       Unimax Art Galleries

In [25]:
# TODO: Create a list of dicts. Each dict represents one product.

"""
Tips:

    To iterate through dataframe:
    
        for _, r in products_all_df.iterrows():
            r = dict(r)
            Access fields like r['productName'], r['productVendor'], ...

"""
products = []
for _, r in products_all_df.iterrows():
    product = {
        "productCode": r["productCode"],
        "productName": r["productName"],
        "productVendor": r["productVendor"],
    }

    # products.append(product)

# print(products)

In [26]:
def insert_products(d):
    mongo_client["w4111"]["products"].drop()
    mongo_client["w4111"]["products"].insert_many(d)


# TODO: Put the name of your list of dicts below
insert_products(products)
print("Successfully inserted product data")

TypeError: documents must be a non-empty list

## MongoDB: Testing

- Run through the following cells
- **Make sure the outputs are completely visible. You shouldn't need to scroll to see the entire output.**
  - You may need to click on the blank section immediately to the left of your output to toggle between scrolling and unscrolling


In [ ]:
import json


def prepr(doc):
    try:
        del doc["_id"]
    except KeyError:
        pass

    def convert_str(d):
        if isinstance(d, dict):
            for k, v in d.items():
                d[k] = convert_str(v)
            return d
        elif isinstance(d, list):
            for i, v in enumerate(d):
                d[i] = convert_str(v)
            return d
        else:
            return str(d)

    convert_str(doc)
    return json.dumps(doc, indent=2)

In [ ]:
res = mongo_client["w4111"]["customers"].aggregate(
    [{"$match": {"customerNumber": 219}}]
)

print(prepr(list(res)[0]))

{
  "customerNumber": "219",
  "customerName": "Boards & Toys Co.",
  "country": "United States",
  "orders": [
    {
      "orderNumber": "10154",
      "orderDate": "2003-10-02",
      "orderContents": [
        {
          "productCode": "S24_3151",
          "quantityOrdered": "31",
          "priceEach": "75.23"
        },
        {
          "productCode": "S700_2610",
          "quantityOrdered": "36",
          "priceEach": "59.27"
        }
      ]
    },
    {
      "orderNumber": "10376",
      "orderDate": "2005-02-08",
      "orderContents": [
        {
          "productCode": "S12_3380",
          "quantityOrdered": "35",
          "priceEach": "98.65"
        }
      ]
    }
  ]
}


In [ ]:
res = mongo_client["w4111"]["customers"].aggregate(
    [{"$match": {"customerNumber": 103}}]
)

print(prepr(list(res)[0]))

{
  "customerNumber": "103",
  "customerName": "Atelier graphique",
  "country": "France",
  "orders": [
    {
      "orderNumber": "10123",
      "orderDate": "2003-05-20",
      "orderContents": [
        {
          "productCode": "S18_1589",
          "quantityOrdered": "26",
          "priceEach": "120.71"
        },
        {
          "productCode": "S18_2870",
          "quantityOrdered": "46",
          "priceEach": "114.84"
        },
        {
          "productCode": "S18_3685",
          "quantityOrdered": "34",
          "priceEach": "117.26"
        },
        {
          "productCode": "S24_1628",
          "quantityOrdered": "50",
          "priceEach": "43.27"
        }
      ]
    },
    {
      "orderNumber": "10298",
      "orderDate": "2004-09-27",
      "orderContents": [
        {
          "productCode": "S10_2016",
          "quantityOrdered": "39",
          "priceEach": "105.86"
        },
        {
          "productCode": "S18_2625",
          "quantityOrd

In [ ]:
res = mongo_client["w4111"]["products"].aggregate(
    [{"$match": {"productCode": "S18_1889"}}]
)

print(prepr(list(res)[0]))

{
  "productCode": "S18_1889",
  "productName": "1948 Porsche 356-A Roadster",
  "productVendor": "Gearbox Collectibles"
}


## Neo4j: All Data


- For the graph database, you will have two types of nodes: `Customer` and `Product`
  - **Make sure to use these exact names**
- An order is represented as a relationship from the `Customer` node to the `Product` node. The type of the relationship should be `ORDERED`.

- You will use `customer_orders_limit_df` to create your graph


In [ ]:
%%sql

customer_orders_limit <<

select
    c.customerNumber, c.customerName, c.country, o.orderNumber,
    o.orderDate, od.productCode, p.productName, p.productVendor,
    od.quantityOrdered, od.priceEach
from classicmodels.orders o
    join classicmodels.customers c using (customerNumber)
    join classicmodels.orderdetails od using (orderNumber)
    join classicmodels.products p using (productCode)
where od.quantityOrdered > 49;

 * mysql+pymysql://root:***@localhost
139 rows affected.
Returning data to local variable customer_orders_limit


In [ ]:
customer_orders_limit_df = customer_orders_limit.DataFrame()
customer_orders_limit_df.head(10)

customerNumber                  customerName         country  orderNumber  \
0             363  Online Diecast Creations Co.   United States        10100   
1             145      Danish Wholesale Imports         Denmark        10105   
2             145      Danish Wholesale Imports         Denmark        10105   
3             278                 Rovelli Gifts           Italy        10106   
4             124  Mini Gifts Distributors Ltd.   United States        10113   
5             148       Dragon Souveniers, Ltd.       Singapore        10117   
6             353            Reims Collectables          France        10121   
7             103             Atelier graphique          France        10123   
8             458    Corrida Auto Replicas, Ltd           Spain        10126   
9             324      Stylish Desk Decors, Co.  United Kingdom        10129   

    orderDate productCode                productName  \
0  2003-01-06    S18_2248         1911 Ford Town Car   
1  2003-02-11    S10_4757        1972 Alfa Romeo GTA   
2  2003-02-11    S24_3816   1940 Ford Delivery Sedan   
3  2003-02-17    S24_3949   Corsair F4U ( Bird Cage)   
4  2003-03-26    S18_4668    1939 Cadillac Limousine   
5  2003-04-16    S72_3212                 Pont Yacht   
6  2003-05-07    S12_2823           2002 Suzuki XREO   
7  2003-05-20    S24_1628  1966 Shelby Cobra 427 S/C   
8  2003-05-28    S18_4600           1940s Ford truck   
9  2003-06-12    S24_3816   1940 Ford Delivery Sedan   

              productVendor  quantityOrdered priceEach  
0   Motor City Art Classics               50     55.09  
1   Motor City Art Classics               50    127.84  
2  Carousel DieCast Legends               50     75.47  
3       Second Gear Diecast               50     55.96  
4       Studio M Art Models               50     43.27  
5      Unimax Art Galleries               50     52.42  
6      Unimax Art Galleries               50    126.52  
7  Carousel DieCast Legends               50     43.27  
8   Motor City Art Classics               50    102.92  
9  Carousel DieCast Legends               50     76.31

- Below is an example of how a customer and their orders are stored in MySQL, and how the graph should look like in Neo4j
  - Note that the same order may be represented as many relationships since one order could contain many products
- The `Customer` nodes should have the following attributes:
  - `customerNumber`
  - `customerName`
  - `country`
- The `Product` nodes should have the following attributes:
  - `productCode`
  - `productName`
  - `productVendor`
- The `ORDERED` relationships should have the following attributes:
  - `orderNumber`
  - `orderDate`
  - `quantityOrdered`
  - `priceEach`

**MySQL relation:**

|     | customerNumber | customerName              | country | orderNumber | orderDate  | productCode | productName                             | productVendor             | quantityOrdered | priceEach |
| --: | -------------: | :------------------------ | :------ | ----------: | :--------- | :---------- | :-------------------------------------- | :------------------------ | --------------: | --------: |
|   0 |            450 | The Sharp Gifts Warehouse | USA     |       10250 | 2004-05-11 | S32_4289    | 1928 Ford Phaeton Deluxe                | Highway 66 Mini Classics  |              50 |      62.6 |
|   1 |            450 | The Sharp Gifts Warehouse | USA     |       10257 | 2004-06-14 | S18_2949    | 1913 Ford Model T Speedster             | Carousel DieCast Legends  |              50 |     92.19 |
|   2 |            450 | The Sharp Gifts Warehouse | USA     |       10400 | 2005-04-01 | S10_4757    | 1972 Alfa Romeo GTA                     | Motor City Art Classics   |              64 |    134.64 |
|   3 |            450 | The Sharp Gifts Warehouse | USA     |       10400 | 2005-04-01 | S18_3856    | 1941 Chevrolet Special Deluxe Cabriolet | Exoto Designs             |              58 |     88.93 |
|   4 |            450 | The Sharp Gifts Warehouse | USA     |       10407 | 2005-04-22 | S18_1589    | 1965 Aston Martin DB5                   | Classic Metal Creations   |              59 |    114.48 |
|   5 |            450 | The Sharp Gifts Warehouse | USA     |       10407 | 2005-04-22 | S18_1749    | 1917 Grand Touring Sedan                | Welly Diecast Productions |              76 |     141.1 |
|   6 |            450 | The Sharp Gifts Warehouse | USA     |       10407 | 2005-04-22 | S18_4933    | 1957 Ford Thunderbird                   | Studio M Art Models       |              66 |     64.14 |
|   7 |            450 | The Sharp Gifts Warehouse | USA     |       10407 | 2005-04-22 | S24_1628    | 1966 Shelby Cobra 427 S/C               | Carousel DieCast Legends  |              64 |     45.78 |
|   8 |            450 | The Sharp Gifts Warehouse | USA     |       10407 | 2005-04-22 | S24_2766    | 1949 Jaguar XK 120                      | Classic Metal Creations   |              76 |     81.78 |
|   9 |            450 | The Sharp Gifts Warehouse | USA     |       10407 | 2005-04-22 | S24_2887    | 1952 Citroen-15CV                       | Exoto Designs             |              59 |     98.65 |

**Neo4j graph:**

<img src="./Neo4j-example.png">


In [ ]:
# Deletes all customers and products (and their relationships).
# Feel free to run this as many times as you want to reset your data.
_ = graph.execute_query(
    """
    match (c:Customer), (p:Product)
    detach delete c, p
"""
)

In [ ]:
# TODO: Write and execute queries to create nodes and relationships

"""
Tips:

    To iterate through dataframe:
    
        for _, r in customer_orders_limit_df.iterrows():
            r = dict(r)
            Access fields like r['customerName'], r['country'], ...
            
    
    The priceEach field are stored as a Decimal object in the dataframe. This type is not 
    compatible with the neo4j API. You can convert it to a string by calling str(r['priceEach']). 


    You should call graph.execute_query to execute your queries. This method takes in a second
    optional argument, a dict. This allows you to do query parameters. For instance, to execute
    the query in the screenshot above, you could run
    
        graph.execute_query(
            "match (c:Customer { customerNumber: $custNum })-[:ORDERED]->(p:Product) return c, p",
            { "custNum": 450 }
        )

"""
# type conversion
customer_orders_limit_df["priceEach"] = customer_orders_limit_df["priceEach"].apply(str)

# Step 1: Create Customer Nodes
customer_query = """
UNWIND $customers AS customer
MERGE (c:Customer {customerNumber: customer.customerNumber})
ON CREATE SET c.customerName = customer.customerName, c.country = customer.country
"""

# Step 2: Create Product Nodes
product_query = """
UNWIND $products AS product
MERGE (p:Product {productCode: product.productCode})
ON CREATE SET p.productName = product.productName, p.productVendor = product.productVendor
"""

# Step 3: Create ORDERED Relationships
relationship_query = """
UNWIND $orders AS data
MATCH (c:Customer {customerNumber: data.customerNumber})
MATCH (p:Product {productCode: data.productCode})
MERGE (c)-[r:ORDERED]->(p)
ON CREATE SET r.orderNumber = data.orderNumber, r.orderDate = data.orderDate,
              r.quantityOrdered = data.quantityOrdered, r.priceEach = data.priceEach
"""

unique_customers = (
    customer_orders_limit_df[["customerNumber", "customerName", "country"]]
    .drop_duplicates()
    .to_dict("records")
)
unique_products = (
    customer_orders_limit_df[["productCode", "productName", "productVendor"]]
    .drop_duplicates()
    .to_dict("records")
)
orders = customer_orders_limit_df.to_dict("records")


# Execute queries
def create_graph(database):
    database.execute_query(customer_query, {"customers": unique_customers})
    database.execute_query(product_query, {"products": unique_products})
    database.execute_query(relationship_query, {"orders": orders})


create_graph(graph)

## Neo4j: Testing

- Run through the following cells
- **Make sure the outputs are fully visible**


In [ ]:
res = []

for r in graph.execute_query(
    """
    match (c:Customer { customerNumber: 412 })-[o:ORDERED]->(p:Product)
    return c, o, p
"""
).records:
    res.append(dict(r["c"]) | dict(r["o"]) | dict(r["p"]))

pandas.DataFrame(res)

country  customerNumber                   customerName  orderNumber  \
0  New Zealand             412  Extreme Desk Decorations, Ltd        10418   
1  New Zealand             412  Extreme Desk Decorations, Ltd        10234   
2  New Zealand             412  Extreme Desk Decorations, Ltd        10268   
3  New Zealand             412  Extreme Desk Decorations, Ltd        10418   

   quantityOrdered   orderDate priceEach productCode  \
0               52  2005-05-16     64.41    S24_2360   
1               50  2004-03-30    146.65    S18_1662   
2               50  2004-07-12    124.59    S18_2325   
3               50  2005-05-16    100.01    S32_4485   

                   productName             productVendor  
0      1982 Ducati 900 Monster  Highway 66 Mini Classics  
1  1980s Black Hawk Helicopter         Red Start Diecast  
2    1932 Model A Ford J-Coupe     Autoart Studio Design  
3    1974 Ducati 350 Mk3 Desmo       Second Gear Diecast

In [ ]:
res = []

for r in graph.execute_query(
    """
    match (c:Customer)-[o:ORDERED]->(p:Product { productCode: 'S12_2823' })
    return c, o, p
"""
).records:
    res.append(dict(r["c"]) | dict(r["o"]) | dict(r["p"]))

pandas.DataFrame(res)

country  customerNumber           customerName  orderNumber  \
0          France             353     Reims Collectables        10121   
1         Austria             382  Salzburg Collectables        10341   
2  United Kingdom             201  UK Collectables, Ltd.        10403   

   quantityOrdered   orderDate priceEach productCode       productName  \
0               50  2003-05-07    126.52    S12_2823  2002 Suzuki XREO   
1               55  2004-11-24    120.50    S12_2823  2002 Suzuki XREO   
2               66  2005-04-08    122.00    S12_2823  2002 Suzuki XREO   

          productVendor  
0  Unimax Art Galleries  
1  Unimax Art Galleries  
2  Unimax Art Galleries

In [ ]:
res = []

for r in graph.execute_query(
    """
    match (c:Customer)-[o:ORDERED { quantityOrdered: 60 }]->(p:Product)
    return c, o, p
"""
).records:
    res.append(dict(r["c"]) | dict(r["o"]) | dict(r["p"]))

pandas.DataFrame(res)

country  customerNumber               customerName  orderNumber  \
0          Spain             141     Euro+ Shopping Channel        10412   
1  United States             362          Gifts4AllAges.com        10414   
2      Australia             282  Souveniers And Things Co.        10420   

   quantityOrdered   orderDate priceEach productCode  \
0               60  2005-05-03    157.49    S18_3232   
1               60  2005-05-06     72.58    S24_3151   
2               60  2005-05-29     60.26    S24_1046   

                        productName         productVendor  
0       1992 Ferrari 360 Spider red  Unimax Art Galleries  
1  1912 Ford Model T Delivery Wagon       Min Lin Diecast  
2        1970 Chevy Chevelle SS 454  Unimax Art Galleries

In [ ]:
res = []

for r in graph.execute_query(
    """
    match (n:Customer | Product) 
    return labels(n) as type, count(*) as count 
    union 
    match ()-[r:ORDERED]->() 
    return type(r) as type, count(*) as count
"""
).records:
    res.append(dict(r))

pandas.DataFrame(res)

type  count
0  [Customer]     57
1   [Product]     84
2     ORDERED    138